test/validation/api/ml/ml.c

```diff
@@ -0,0 +1,501 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2023 Nokia
+ */
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+#include "odp_cunit_common.h"
+
+#define UAREA     0xaa
+#define NUM_COMPL 10u
+#define COMPL_POOL_NAME "ML compl pool"
+
+typedef struct global_t {
+	int disabled;
+	uint32_t num_compl;
+	odp_ml_capability_t ml_capa;
+} global_t;
+
+typedef struct {
+	uint32_t count;
+	uint8_t mark[NUM_COMPL];
+} uarea_init_t;
+
+static global_t global;
+
+static int ml_suite_init(void)
+{
+	memset(&global, 0, sizeof(global_t));
+
+	if (odp_ml_capability(&global.ml_capa)) {
+		ODPH_ERR("ML capability failed\n");
+		return -1;
+	}
+
+	if (global.ml_capa.max_models == 0) {
+		global.disabled = 1;
+		ODPH_DBG("ML test disabled\n");
+		return 0;
+	}
+
+	global.num_compl = ODPH_MIN(NUM_COMPL, global.ml_capa.pool.max_num);
+
+	return 0;
+}
+
+static int check_ml_support(void)
+{
+	if (global.disabled)
+		return ODP_TEST_INACTIVE;
+
+	return ODP_TEST_ACTIVE;
+}
+
+static void test_ml_capability(void)
+{
+	odp_ml_capability_t ml_capa;
+
+	memset(&ml_capa, 0, sizeof(odp_ml_capability_t));
+	CU_ASSERT_EQUAL(odp_ml_capability(&ml_capa), 0);
+
+	if (ml_capa.max_models == 0)
+		return;
+
+	CU_ASSERT(ml_capa.max_model_size > 0);
+	CU_ASSERT(ml_capa.max_models_loaded > 0);
+	CU_ASSERT(ml_capa.max_inputs > 0);
+	CU_ASSERT(ml_capa.max_outputs > 0);
+	CU_ASSERT(ml_capa.max_compl_id > 0);
+	CU_ASSERT(ml_capa.max_segs_per_input > 0);
+	CU_ASSERT(ml_capa.max_segs_per_output > 0);
+	CU_ASSERT(ml_capa.min_input_align > 0);
+	CU_ASSERT(ml_capa.min_output_align > 0);
+
+	if ((ml_capa.load.compl_mode_mask | ml_capa.run.compl_mode_mask) &
+	    ODP_ML_COMPL_MODE_EVENT) {
+		odp_pool_capability_t pool_capa;
+
+		CU_ASSERT_FATAL(odp_pool_capability(&pool_capa) == 0);
+
+		CU_ASSERT(ml_capa.pool.max_pools > 0);
+		CU_ASSERT(ml_capa.pool.max_pools <= pool_capa.max_pools);
+		CU_ASSERT(ml_capa.pool.max_num > 0);
+		CU_ASSERT(ml_capa.pool.max_cache_size >= ml_capa.pool.min_cache_size);
+	}
+
+	CU_ASSERT(ml_capa.load.compl_mode_mask);
+	CU_ASSERT(ml_capa.run.compl_mode_mask);
+
+	if (ml_capa.load.compl_mode_mask & ODP_ML_COMPL_MODE_POLL ||
+	    ml_capa.run.compl_mode_mask & ODP_ML_COMPL_MODE_POLL)
+		CU_ASSERT(ml_capa.max_compl_id > 0);
```

Comment:
This is max ID (not num IDs). In theory an implementation could have only one possible ID (== 0). So, I'd remove the check.